### PR TITLE
Fixing some bugs in auto_regressive_seq_decoder

### PR DESCRIPTION
**Pull request by nicola-decao**
Fixing for https://github.com/allenai/allennlp/issues/3407 and https://github.com/allenai/allennlp/issues/3426.

In https://github.com/allenai/allennlp/blob/f3083c8fb9150f07e3ca98bb3ea9368a081df028/allennlp/modules/seq2seq_decoders/auto_regressive_seq_decoder.py#L433

target_tokens["tokens"] is a torch.Tensor so its elements do not have the attribute text.

### CODE REVIEW

**Comment by brendan-ai2** [comment]
Hi @nicola-decao, thanks for this PR! Any chance you could take a stab at writing a test for this?

File: allennlp/modules/seq2seq_decoders/auto_regressive_seq_decoder.py [hunk 1]
```diff
@@ -442,10 +442,17 @@ def post_process(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.
...
         predicted_indices = output_dict["predictions"]
-        if not isinstance(predicted_indices, numpy.ndarray):
-            predicted_indices = predicted_indices.detach().cpu().numpy()
-        all_predicted_tokens = []
-        for indices in predicted_indices:
+        all_predicted_tokens = self.indeces_to_tokens(predicted_indices)
+        output_dict["predicted_tokens"] = all_predicted_tokens
+        return output_dict
+
+    def indeces_to_tokens(self, batch_indeces: numpy.ndarray) -> List[List[str]]:
```
Comment (sai-prasanna) [inline allennlp/modules/seq2seq_decoders/auto_regressive_seq_decoder.py]: Small typo, "indeces" -> "indices" or "indexes".

File: allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        target_embedding_dim=decoder_inout_dim,
+        feedforward_hidden_dim=20,
+        num_layers=2,
+        num_attention_heads=4,
+    )
+
+    return vocab, decoder_net
+
+
+class TestAutoRegressiveSeqDecoder(AllenNlpTestCase):
```
Comment (sai-prasanna) [inline allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py]: Can you add a test for whether the metrics (`self._token_based_metric` and `self._tensor_based_metric`) get called when present?

**Comment by sai-prasanna** [comment]
@brendan-ai2 Can you take a look? I think required tests were added by @nicola-decao

**Comment by nicola-decao** [comment]
@brendan-ai2 the test which fails seems unrelated to the edits. I cannot understand why it fails. As I say in https://github.com/allenai/allennlp/pull/3464 there might be a problem in the master branch. Both my pull request, #3461, and #3453 fails the same test but do not seem correlated.

**Comment by brendan-ai2** [comment]
@nicola-decao, sorry about those failures. Those definitely aren't your fault. We updated Spacy versions and it created some churn in our tokenization. We're hoping to have a fix in master shortly.

**Comment by nicola-decao** [comment]
> @nicola-decao, sorry about those failures. Those definitely aren't your fault. We updated Spacy versions and it created some churn in our tokenization. We're hoping to have a fix in master shortly.

Yep, I saw https://github.com/allenai/allennlp/pull/3468 . Thanks for letting me know :)

File: allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            Embedding(vocab.get_vocab_size(), decoder_inout_dim),
+            tensor_based_metric=BLEU(),
+            token_based_metric=CustomSquadEmAndF1(),
+        ).eval()
+
+        encoded_state = torch.randn(batch_size, time_steps, decoder_inout_dim)
+        source_mask = torch.ones(batch_size, time_steps).long()
+        target_tokens = {"tokens": torch.ones(batch_size, time_steps).long()}
+        source_mask[0, 1:] = 0
+        encoder_out = {"source_mask": source_mask, "encoder_outputs": encoded_state}
```
Comment (brendan-ai2) [inline allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py]: Could you clarify why the asserted metric are fixed even though the encoded state is random? Looks great otherwise! Thanks for the tests!

Comment (nicola-decao) [reply allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py]: The encoder state is not random. `prepare_environment(Params({}))` is setting all the seeds to a fixed value so `torch.randn` is returning always the same value and that is why the test passes when checking the BLEU score.

Comment (brendan-ai2) [reply allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py]: Ah, right. Thanks! Could you add that as a comment?

Comment (nicola-decao) [reply allennlp/tests/modules/seq2seq_decoders/auto_regressive_seq_decoder_test.py]: Done! ;)

**Comment by brendan-ai2** [comment]
Great, thanks again!
